app/interactors/update_task.rb

```diff
@@ -2,19 +2,43 @@ class UpdateTask
   include Interactor
   include Errors
 
-  def call
+  before do
     return object_not_found!(:task) unless context.task
 
-    swap_position(context.position) if context.params[:position].present?
-    return context if context.task.update(context.params)
+    validate_position if context.params[:position].present?
+  end
+
+  after { check_tasks_complete if context.params[:done].present? }
 
-    object_invalid!(:task)
+  def call
+    return unless context.success?
+
+    change_position(context.params[:position].to_i) if context.params[:position].present?
+    return object_invalid!(:task) unless context.task.update(context.params.except(:position))
+
+    context.data = {}
   end
 
   private
 
-  def swap_position(position)
-    another_task = Task.find_by(project: context.task.project, position: position)
-    another_task&.update(position: context.task.position)
+  def change_position(position)
+    context.task.insert_at(position)
+  end
+
+  def all_tasks_complete(tasks)
+    tasks.each { |t| return false unless t.done }
+
+    true
+  end
```

Comment:
Lets image that your project has 100_000_000 tasks
In this case, you get 100_000_000 objects from DB, store them in the memory and try to check if any of them have `done: false`
You need to use the ActiveRecord method for this because PostgreSQL is much more powerful for this operation that Ruby. Please find a solution for optimizing this query.